File: src/main/java/com/github/pareronia/i18n_puzzles/Puzzle2025_03.java
package com.github.pareronia.i18n_puzzles;

import java.util.List;

import com.github.pareronia.i18n_puzzles.common.Sample;
import com.github.pareronia.i18n_puzzles.common.Samples;
import com.github.pareronia.i18n_puzzles.common.SolutionBase;
import com.github.pareronia.i18n_puzzles.common.Utils;

public class Puzzle2025_03 extends SolutionBase<Integer>{

    private Puzzle2025_03(final boolean debug) {
        super(debug);
    }

    public static Puzzle2025_03 create() {
        return new Puzzle2025_03(false);
    }

    public static Puzzle2025_03 createDebug() {
        return new Puzzle2025_03(true);
    }

    private boolean isValid(final String password) {
        final long count = password.chars().count();
        if (count < 4 || count > 12) {
            return false;
        }
        final int bits = Utils.asCharacterStream(password)
                .mapToInt(ch -> {
                    int ans = 0;
                    if (Character.isDigit(ch)) {
                        ans |= 0b0001;
                    }
                    if (Character.isUpperCase(ch)) {
                        ans |= 0b0010;
                    }
                    if (Character.isLowerCase(ch)) {
                        ans |= 0b0100;
                    }
                    if (ch.charValue() > 127) {
                        ans |= 0b1000;
                    }
                    return ans;
                })
                .reduce((a, b) -> a | b).getAsInt();
        return bits == 0b1111;
    }
    
    @Override
    public Integer solve(final List<String> input) {
        return (int) input.stream().filter(this::isValid).count();
    }

    @Override
    @Samples({
        @Sample(input = TEST, expected = "2")
    })
    public void samples() { }

    public static void main(final String[] args) throws Exception {
        Puzzle2025_03.create().run();
    }

    private static final String TEST = """
            d9Ō
            uwI.E9GvrnWļbzO
            ž-2á
            Ģ952W*F4
            ?O6JQf
            xi~Rťfsa
            r_j4XcHŔB
            71äĜ3
            """;
}

File: src/main/python/Puzzle2025_03.py
import operator
import sys
from functools import reduce

from i18n_puzzles.common import InputData
from i18n_puzzles.common import SolutionBase
from i18n_puzzles.common import i18n_puzzles_samples

TEST = """\
d9Ō
uwI.E9GvrnWļbzO
ž-2á
Ģ952W*F4
?O6JQf
xi~Rťfsa
r_j4XcHŔB
71äĜ3
"""

Output = int


class Solution(SolutionBase[Output]):
    def solve(self, input: InputData) -> Output:
        def is_valid(line: str) -> bool:
            return (
                4 <= len(line) <= 12
                and reduce(
                    operator.or_,
                    map(
                        lambda ch: ch.isnumeric() * 0b0001
                        + ch.isupper() * 0b0010
                        + ch.islower() * 0b0100
                        + (ord(ch) > 127) * 0b1000,
                        (ch for ch in line),
                    ),
                )
                == 0b1111
            )

        return sum(is_valid(line) for line in input)

    @i18n_puzzles_samples(((TEST, 2),))
    def samples(self) -> None:
        pass


solution = Solution(3)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
